Core/Systems/CutsceneSystem/CutscenePlayer.cs

```diff
@@ -0,0 +1,201 @@
+﻿using NetEasy;
+using StarlightRiver.Content.GUI;
+using StarlightRiver.Core.Loaders.UILoading;
+using System;
+using Terraria.ID;
+
+namespace StarlightRiver.Core.Systems.CutsceneSystem
+{
+	internal class CutscenePlayer : ModPlayer
+	{
+		public Cutscene activeCutscene;
+
+		public int fadeTimer;
+
+		public bool InCutscene => activeCutscene != null;
+
+		/// <summary>
+		/// Sets a given cutscene type as active for this player. This activates cutscene mode.
+		/// </summary>
+		/// <typeparam name="T"></typeparam>
+		public void SetActiveCutscene<T>() where T : Cutscene
+		{
+			activeCutscene?.EndCutscene(Player); // end current cutscene to begin another
+			activeCutscene = (Cutscene)Activator.CreateInstance(typeof(T).Assembly.FullName, typeof(T).FullName).Unwrap();
+
+			if (Main.netMode == NetmodeID.MultiplayerClient)
+			{
+				var packet = new CutscenePacket((byte)Player.whoAmI, typeof(T).FullName, true);
+				packet.Send(-1, Player.whoAmI, false);
+			}
+		}
+
+		/// <summary>
+		/// Sets a given cutscene type as active for this player. This activates cutscene mode.
+		/// 
+		/// This version should typically only be called from net sync packets
+		/// </summary>
+		/// <param name="fullName">The fully qualified name of the cutscene type</param>
+		public void SetActiveCutscene(string fullName)
+		{
+			Mod.Logger.Info($"Activating cutscene {fullName} for {Player.name}");
+			activeCutscene?.EndCutscene(Player); // end current cutscene to begin another
+			activeCutscene = (Cutscene)Activator.CreateInstance(GetType().Assembly.FullName, fullName).Unwrap();
+		}
+
+		/// <summary>
+		/// Deactivates a cutscene if one is active
+		/// </summary>
+		public void DeactivateCutscene()
+		{
+			Mod.Logger.Info($"Ending cutscene for {Player.name}");
+			activeCutscene?.EndCutscene(Player);
+			activeCutscene = null;
+
+			if (Main.netMode == NetmodeID.MultiplayerClient)
+			{
+				var packet = new CutscenePacket((byte)Player.whoAmI, "", false);
+				packet.Send(-1, Player.whoAmI, false);
+			}
+		}
+
+		public override void PreUpdate()
+		{
+			if (InCutscene)
+			{
+				activeCutscene?.UpdateCutscene(Player);
+
+				if (fadeTimer < 60)
+					fadeTimer++;
+			}
+			else
+			{
+				if (fadeTimer > 0)
+					fadeTimer--;
+			}
+		}
+
+		public override void SetControls()
```

Comment:
not necessary but controlsPlayer could have its packets not sent for frames in a cutscene too for a minor reduction in packets needed

Reply:
Perhaps? Seems like this could slowly start to enter spaghetti territory if we're checking against random different things, especially when the overlap is likely to be very minimal over the course of gameplay